File: packages/react-native-fipa/android/src/main/java/com/deviceattestation/FirstPartyLocalOrigin.kt
package com.deviceattestation

import java.util.Locale

/** Receives a parsed hostname, never an authority or an unparsed URL. */
internal object FirstPartyLocalOrigin {
    fun contains(hostname: String): Boolean {
        val host = hostname.lowercase(Locale.ROOT)
        if (host in setOf("127.0.0.1", "::1", "[::1]")) return true
        val name = host.removeSuffix(".")
        if (name.length > 253) return false
        val labels = name.split('.')
        return labels.last() == "localhost" &&
            labels.all { it.matches(Regex("[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?")) }
    }
}

File: packages/react-native-fipa/ios/FirstPartyLocalOrigin.swift
import Foundation

/// Receives a parsed hostname, never an authority or an unparsed URL.
enum FirstPartyLocalOrigin {
  static func contains(_ hostname: String) -> Bool {
    let host = hostname.lowercased()
    if ["127.0.0.1", "::1", "[::1]"].contains(host) { return true }
    let name = host.hasSuffix(".") ? String(host.dropLast()) : host
    guard name.utf8.count <= 253 else { return false }
    let labels = name.split(separator: ".", omittingEmptySubsequences: false)
    return labels.last == "localhost" && labels.allSatisfy {
      $0.range(of: "^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?$", options: .regularExpression) != nil
    }
  }
}
